scipy/stats/_mannwhitneyu.py

```diff
@@ -0,0 +1,367 @@
+        raise ValueError('`axis` must be an integer.')
+
+    methods = {"asymptotic", "exact"}
+    method = method.lower()
+    if method not in methods:
+        raise ValueError(f'`method` must be one of {methods}.')
+
+    return x, y, use_continuity, alternative, axis_int, method
+
+
+mwu_result = make_dataclass("MannWhitneyUResult", ("statistic", "pvalue"))
+# Using `nametuple` for now to pass existing mannwhitneyu tests without
+# modification
+MannwhitneyuResult = namedtuple('MannwhitneyuResult', ('statistic', 'pvalue'))
+
+
+def mannwhitneyu2(x, y, use_continuity=True, alternative="two-sided",
+                  axis=0, method="asymptotic"):
+    r'''Perform the Mann-Whitney U rank test on two independent samples.
+
+    The Mann-Whitney U test is a nonparametric test of the null hypothesis
+    that the underlying distribution of sample `x` is neither stochastically
+    greater nor stochastically less than the uderlying distribution of sample
+    `y`.
+
+    Parameters
+    ----------
+    x, y : array-like
+        N-d arrays of samples. The arrays must be broadcastable except along
+        the dimension given by `axis`.
+    use_continuity : bool, optional
+            Whether a continuity correction (1/2) should be applied.
+            Default is True when `method` is ``'asymptotic'``; has no effect
+            otherwise.
+    alternative : {'two-sided', 'less', 'greater'}, optional
+        Defines the alternative hypothesis.
+        The following options are available:
+
+        * 'two-sided' (default)
+        * 'less': one-sided
+        * 'greater': one-sided
+
+    axis : int, optional
+        Axis along which to perform the test. Default is 0.
+    method : {'asymptotic', 'exact'}, optional
+        Selects the method used to calculate the *p*-value. The following
+        options are available.
+
+        * ``'asymptotic'``: (default) compares the standardized test statistic
+          against the normal distribution, correcting for ties.
+        * ``'exact'``: computes the exact *p*-value by comparing the observed
+          :math:`U` statistic against the exact distribution of the :math:`U`
+          statistic under the null hypothesis. No correction is made for ties.
+
+    Returns
+    -------
+    res : MannwhitneyuResult
+        An object containing attributes:
+
+        statistic : float
+            The Mann-Whitney U statistic corresponding with sample `x`. See
+            Notes for the test statistic corresponding with sample `y`.
+        pvalue : float
+            The associated *p*-value for the chosen `alternative`.
+
+    Notes
+    -----
+    If ``U1`` is the statistic corresponding with sample `x`, then the
+    statistic corresponding with sample `y` is
+    `U2 = `x.shape[axis] * y.shape[axis] - U1``.
+
+    `mannwhitneyu2` is for independent samples. For related samples,
+    consider `scipy.stats.wilcoxon`.
+
+    `method` ``'exact'`` is recommended when there are no ties and when either
```

Comment:
this could be a criterion if method=auto is used

Reply:
Done except need to switch to asymptotic in case of ties.

Reply:
Done.